lib/puppet/provider/exec.rb

```diff
@@ -26,6 +26,11 @@ def environment
         warning _("Overriding environment setting '%{var}' with '%{value}'") % { var: var, value: value }
       end
 
+      if value.nil? || value.empty?
+        msg = _("Empty environment setting '%{var}'") % {var: var}
+        Puppet.warn_once('undefined_variables', "empty_env_var_#{var}", msg, __FILE__, __LINE__)
```

Comment:
It looks like `__FILE__` and `__LINE__` are returning where we are in the ruby code, rather than a manifest
```
[0] Melissa@bubba:/Users/Melissa/puppet:(PUP-1209)$ be ./bin/puppet apply -e "exec{'foobar': command => '/usr/bin/true', environment => ['FOO=','BAR=','FOO=']}"
Notice: Compiled catalog for bubba.local in environment production in 0.07 seconds
Warning: Empty environment setting 'FOO'
   (file: /Users/Melissa/puppet/lib/puppet/provider/exec.rb, line: 31)
Warning: Empty environment setting 'BAR'
   (file: /Users/Melissa/puppet/lib/puppet/provider/exec.rb, line: 31)
Warning: Exec[foobar](provider=posix): Overriding environment setting 'FOO' with ''
Notice: /Stage[main]/Main/Exec[foobar]/returns: executed successfully
Notice: Applied catalog in 0.03 seconds
[0] Melissa@bubba:/Users/Melissa/puppet:(PUP-1209)$ be ./bin/puppet apply test.pp
Notice: Compiled catalog for bubba.local in environment production in 0.07 seconds
Warning: Empty environment setting 'FOO'
   (file: /Users/Melissa/puppet/lib/puppet/provider/exec.rb, line: 31)
Warning: Empty environment setting 'BAR'
   (file: /Users/Melissa/puppet/lib/puppet/provider/exec.rb, line: 31)
Warning: Exec[foobar](provider=posix): Overriding environment setting 'FOO' with ''
Notice: /Stage[main]/Main/Exec[foobar]/returns: executed successfully
Notice: Applied catalog in 0.02 seconds
```

Rather than ` __FILE__` and `__LINE__`, use `resource.file` and `resource.line`, which will give us more useful info on where the unset var is
```
[0] Melissa@bubba:/Users/Melissa/puppet:(PUP-1209)$ be ./bin/puppet apply test.pp
Notice: Compiled catalog for bubba.local in environment production in 0.07 seconds
Warning: Empty environment setting 'FOO'
   (file: /Users/Melissa/puppet/test.pp, line: 1)
Warning: Empty environment setting 'BAR'
   (file: /Users/Melissa/puppet/test.pp, line: 1)
Warning: Exec[foobar](provider=posix): Overriding environment setting 'FOO' with ''
Notice: /Stage[main]/Main/Exec[foobar]/returns: executed successfully
Notice: Applied catalog in 0.02 seconds
[0] Melissa@bubba:/Users/Melissa/puppet:(PUP-1209)$ be ./bin/puppet apply -e "exec{'foobar': command => '/usr/bin/true', environment => ['FOO=','BAR=','FOO=']}"
Notice: Compiled catalog for bubba.local in environment production in 0.07 seconds
Warning: Empty environment setting 'FOO'
   (line: 1)
Warning: Empty environment setting 'BAR'
   (line: 1)
Warning: Exec[foobar](provider=posix): Overriding environment setting 'FOO' with ''
Notice: /Stage[main]/Main/Exec[foobar]/returns: executed successfully
Notice: Applied catalog in 0.03 seconds
```

```suggestion
        Puppet.warn_once('undefined_variables', "empty_env_var_#{var}", msg, resource.file, resource.line)
```

Reply:
I'm fine with your suggestion, but I not sure what to do with the tests.

They're currently failing with this output:

```
  1) Puppet::Type::Exec::ProviderPosix#run should warn when setting an empty environment variable
     Failure/Error: expect(@logs.map(&:to_s).join).to match(/Empty environment setting 'WHATEVER'\n\s+\(file: .+, line: \d+\)/m)
     
       expected "Empty environment setting 'WHATEVER'\n   (file & line not available)" to match /Empty environment setting 'WHATEVER'\n\s+\(file: .+, line: \d+\)/m
       Diff:
       @@ -1,2 +1,3 @@
       -/Empty environment setting 'WHATEVER'\n\s+\(file: .+, line: \d+\)/m
       +Empty environment setting 'WHATEVER'
       +   (file & line not available)
       
     # ./spec/unit/provider/exec/posix_spec.rb:128:in `block (3 levels) in <top (required)>'
```

I don't know what would be better:

* match `(file & line not available)` even if it will only appear in the test environment
* match the warning message and anything that comes after (e.g. `/Empty environment setting 'WHATEVER'.+/m`.

What do you think?

Reply:
I would have at least two test cases for this. One where file and line aren't available and one where they are. You can set file and line with `provider.resource.line = '2'` and `provider.resource.file = '/not/a/file'` respectively

Reply:
Isn't that out of the scope of the `exec` resource tests? It feels to me that doing this is like testing `warn_once`.

I updated the PR anyway, so we can move forward if you think that's not an issue.